### PR TITLE
[MODUSERS-354] Consume Primary Affiliation Created event and store in local persistence

### PR DESCRIPTION
**Pull request by Abdulkhakimov**
## Purpose
We need to implement a consumer of CONSORTIUM_PRIMARY_AFFILIATION_CREATED event that is produced by mod-consortia module. This event will include userId, username, tenantId in the payload and these data should be stored in user_tenant table in mod-login.

## Approach
- Implemented kafka consumer for event: CONSORTIUM_PRIMARY_AFFILIATION_CREATED
- Implemented endpoint to retrieve saved user affiliations

## Learning
[MODUSERS-354](https://issues.folio.org/browse/MODUSERS-354)

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -442,8 +458,8 @@
         "departments.item.delete",
         "patron-pin.set",
         "patron-pin.delete",
-        "patron-pin.validate"
-      ]
+        "patron-pin.validate",
+        "user-tenants.collection.get"      ]
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply descriptors/ModuleDescriptor-template.json]: We fixed this formatting issue

File: src/main/java/org/folio/rest/impl/InitAPIs.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    try {
+      deployConsumersVerticles(vertx)
+        .onSuccess(car -> {
+          handler.handle(Future.succeededFuture());
+          logger.info("Consumer Verticles were successfully started");
+        })
+        .onFailure(th -> {
+          handler.handle(Future.failedFuture(th));
+          logger.error("Consumer Verticles were not started", th);
+        });
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply src/main/java/org/folio/rest/impl/InitAPIs.java]: applied improvement, thanks

File: src/main/java/org/folio/rest/impl/InitAPIs.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    int usersConsortiumConsumerInstancesNumber = Integer.parseInt(getPropertyValue("users.consortium.kafka.consumer.instancesNumber", "1"));
+
+    vertx.deployVerticle((ConsortiumEventConsumersVerticle.class.getName()),
+      new DeploymentOptions()
+        .setWorker(true)
+        .setInstances(usersConsortiumConsumerInstancesNumber), consortiaEventsConsumer);
+
+    return GenericCompositeFuture.all(Collections.singletonList(
+      consortiaEventsConsumer.future()));
+  }
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply src/main/java/org/folio/rest/impl/InitAPIs.java]: applied improvement, thanks

File: src/test/java/org/folio/moduserstest/AbstractRestTest.java [hunk 4]
```diff
@@ -84,6 +99,7 @@ public static void beforeAll(Vertx vertx, VertxTestContext context) {
   public static void after(VertxTestContext context) {
     module.purgeModule(okapiHeaders)
       .compose(v -> {
+        kafkaProducer.close();
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply src/test/java/org/folio/moduserstest/AbstractRestTest.java:102]: fixed, thank you

File: src/test/java/org/folio/rest/impl/UserTenantITTest.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  @Test
+  void canCreateUserTenant() {
+    //given
+    String eventPayload = Json.encode(USER_AFFILIATION);
+    //when
+    sendEvent(TENANT_NAME, ConsortiumEventType.CONSORTIUM_PRIMARY_AFFILIATION_CREATED.getTopicName(),
+      USER_AFFILIATION.getId(), eventPayload);
+    awaitHandlingEvent();
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply src/test/java/org/folio/rest/impl/UserTenantITTest.java]: moved duplicated code into new sendAffiliationCreatedEvent() method

File: src/test/java/org/folio/rest/impl/UserTenantITTest.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  @SneakyThrows
+  static void beforeAll() {
+    userTenantClient = new UserTenantClient(okapiUrl, okapiHeaders);
+  }
+
+  @BeforeEach
+  public void beforeEach() {
+    LOAD_SAMPLE_DATA = false;
+    LOAD_REFERENCE_DATA = false;
+  }
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply src/test/java/org/folio/rest/impl/UserTenantITTest.java]: applied this suggestion with 2 other abstract classes

**Review by SinghAdes — APPROVED** [review]

File: src/test/java/org/folio/rest/impl/UserTenantITTest.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+@ExtendWith(VertxExtension.class)
+class UserTenantITTest extends AbstractRestTest {
+
+  private static UserTenantClient userTenantClient;
+  private static final UserTenant USER_AFFILIATION =  new UserTenant()
+    .withId(UUID.randomUUID().toString())
+    .withUserId(UUID.randomUUID().toString())
+    .withUserName("folio_user")
+    .withTenantId("folio_tenant");
```
[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply src/test/java/org/folio/rest/impl/UserTenantITTest.java]: Thanks. updated

**Review by SerhiiNosko — APPROVED** [review]

File: src/test/java/org/folio/rest/InstallUpgradeIT.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * Check the shaded fat uber jar and Dockerfile.
+ *
+ * <p>Test /admin/health.
+ *
+ * <p>Test that logging works.
+ *
+ * <p>Test installation and migration with smoke test.
+ *
```
[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply src/test/java/org/folio/rest/InstallUpgradeIT.java]: Deleted

File: ramls/userTenants.raml [hunk 9]
```diff
@@ -0,0 +1,60 @@
+#%RAML 1.0
+title: User tenants
+baseUri: http://github.com/org/folio/mod-users
+
+documentation:
+  - title: mod-users User tenants API
+    content: This documents the API calls that can be made to query primary users affiliation
```
[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply ramls/userTenants.raml]: Added. thank you

File: ramls/userTenantCollection.json [hunk 10]
```diff
@@ -0,0 +1,23 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "description": "Collection of user tenants",
```
[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply ramls/userTenantCollection.json]: Updated

File: ramls/userTenant.json [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    "userId": {
+      "description": "UUID of the user",
+      "$ref": "raml-util/schemas/uuid.schema"
+    },
+    "userName": {
+      "description": "The user name",
+      "type": "string"
+    },
+    "tenantId": {
+      "description": "UUID of the tenant",
```
[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply ramls/userTenant.json]: Updated

[the remark this answers is not in the document]
Comment (Abdulkhakimov) [reply ramls/userTenant.json]: Updated. Many thanks

**Review by tatsiana-tarhonskaya — APPROVED** [review]

File: src/main/java/org/folio/verticle/consumers/ConsortiumEventsHandler.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+          event.getId(), event.getUserId(), event.getTenantId());
+        result.complete(event.getId());
+      })
+      .onFailure(e -> {
+        if (e instanceof DuplicateEventException) {
+          logger.info("Duplicate user primary affiliation event with consortium id: {} for user id: {} with tenant id: {} received, skipped processing",
+            event.getId(), event.getUserId(), event.getTenantId());
+          result.complete(event.getId());
+        } else {
+          logger.error("Trying to save of user primary affiliation event with consortium id: {} for user id: {} with tenant id: {} has been failed",
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/verticle/consumers/ConsortiumEventsHandler.java:53]: Maybe change to "Attempt to save user primary affilliation event ..." ... "failed"?

**Pull request merged**